src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -287,6 +287,16 @@ private ReadOnlyMemory<byte> ReadChunkFromConnectionBuffer(int maxBytesToRead)
 
                                 if (currentLine.IsEmpty)
                                 {
+                                    // Dispose of the registration and then check whether cancellation has been
+                                    // requested. This is necessary to make determinstic a race condition between
+                                    // cancellation being requested and unregistering from the token.  Otherwise,
+                                    // it's possible cancellation could be requested just before we unregister and
+                                    // we then return a connection to the pool that has been or will be disposed
+                                    // (e.g. if a timer is used and has already queued its callback but the
+                                    // callback hasn't yet run).
+                                    cancellationRegistration.Dispose();
```

Comment:
I wonder if it might make the code a bit cleaner if we changed this so we don't call CompleteResponse here.  Just set state to Done and expect callers to call CompleteResponse and deal with cancellation if necessary.

Reply:
We could, but we'd end up needing to do that check and call in CopyToAsync, ReadAsync, ReadAsyncCore, and DrainAsync... I'm not sure it's any cleaner.